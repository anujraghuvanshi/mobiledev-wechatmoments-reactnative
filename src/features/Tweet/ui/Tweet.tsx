import React, {ReactElement, useState} from 'react';
import {
  Image,
  StyleSheet,
  View,
  Text,
  ViewStyle,
  TextStyle,
} from 'react-native';

import {BasicStyle, ITweet} from '../../../types';
import {TouchableOpacity} from 'react-native-gesture-handler';

interface ITweetProps {
  tweet: ITweet;
}

export function Tweet({tweet}: ITweetProps): ReactElement {
  const [showAllImages, setshowAllImages] = useState<boolean>(
    tweet.images && tweet?.images.length <= 4 ? true : false,
  );
  const [isReadingMore, setisReadingMore] = useState(false);

  return (
    <View testID="tweet-wrapper" style={styles.container}>
      {tweet.error || tweet['unknown error'] ? (
        <Text testID="error-text">
          Tweet with error: {tweet.error || tweet['unknown error']}{' '}
        </Text>
      ) : (
        <>
          <Image
            style={styles.image}
            testID="sender-image"
            source={{
              uri: tweet?.sender?.avatar || '',
              width: 40,
              height: 40,
            }}
          />
          <View style={styles.tweetContainer}>
            <View>
              <Text style={styles.sender}>
                {tweet?.sender?.nick || tweet?.sender?.username}
              </Text>
              {tweet?.content && (
                <View style={styles.contentWrapper}>
                  <Text
                    testID="tweet-content"
                    style={styles.text}
                    numberOfLines={isReadingMore ? undefined : 2}>
                    {tweet.content}
                  </Text>
                  <Text
                    style={styles.readMore}
                    testID="read-more-trigger"
                    onPress={() => {
                      setisReadingMore(!isReadingMore);
                    }}>
                    {isReadingMore ? 'Read Less' : 'Read more'}
                  </Text>
                </View>
              )}
            </View>
            {tweet?.images?.length && (
              <View style={styles.tweetImageContainer}>
                <View
                  testID="tweet-images-wrapper"
                  style={styles.imagesWrapper}>
                  {(showAllImages
                    ? tweet.images
                    : tweet.images.slice(0, 4)
                  ).map((image, index) => (
                    <Image
                      testID="tweet-image"
                      key={`image-${image?.url?.split('/').pop()}-${index}`}
                      style={styles.image}
                      source={{
                        uri: image?.url || '',
                        width: 64,
                        height: 64,
                      }}
                    />
                  ))}
                </View>
                {tweet?.images?.length > 4 && (
                  <TouchableOpacity
                    onPress={() => {
                      setshowAllImages(!showAllImages);
                    }}>
                    <Text>Show All Images</Text>
                  </TouchableOpacity>
                )}
              </View>
            )}
            {tweet.comments && (
              <View testID="tweet-comments-wrapper" style={styles.commentsWrap}>
                {tweet.comments.map((comment, index) => (
                  <View
                    key={`${index}-${comment.sender.username}`}
                    style={styles.comment}>
                    <Text style={styles.commentSender}>
                      {comment.sender.nick}:{' '}
                    </Text>
                    <Text>{comment.content}</Text>
                  </View>
                ))}
              </View>
            )}
          </View>
        </>
      )}
    </View>
  );
}

interface AdditionalStyle {
  imagesWrapper: ViewStyle;
  tweetContainer: ViewStyle;
  sender: TextStyle;
  commentsWrap: ViewStyle;
  comment: ViewStyle;
  commentSender: TextStyle;
  tweetImageContainer: ViewStyle;
  contentWrapper: ViewStyle;
  readMore: TextStyle;
}

const styles: Partial<BasicStyle> & AdditionalStyle = StyleSheet.create<
  Partial<BasicStyle> & AdditionalStyle
>({
  container: {
    flex: 1,
    alignContent: 'space-between',
    backgroundColor: 'white',
    flexDirection: 'row',
    paddingBottom: 8,
    paddingTop: 8,
    paddingRight: 8,
    borderBottomColor: 'lightgrey',
    borderBottomWidth: 1,
  },
  image: {
    marginRight: 16,
    marginBottom: 16,
    backgroundColor: '#e4f0f5',
  },
  imagesWrapper: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    paddingTop: 8,
  },
  tweetContainer: {
    flexShrink: 1,
    flex: 1,
  },
  text: {
    flexWrap: 'wrap',
    flexShrink: 1,
    color: '#a1a1a1',
  },
  sender: {
    color: '#4152c9',
    fontWeight: '600',
  },
  commentsWrap: {
    paddingTop: 8,
    flex: 1,
    flexDirection: 'column',
  },
  comment: {
    flex: 1,
    width: '100%',
    marginBottom: 4,
    flexDirection: 'row',
    backgroundColor: '#f7f7f7',
    padding: 2,
  },
  commentSender: {
    fontWeight: 'bold',
    color: '#000',
  },
  tweetImageContainer: {
    flex: 1,
    flexDirection: 'column',
  },
  contentWrapper: {
    flexDirection: 'column',
  },
  readMore: {
    color: 'blue',
  },
});
