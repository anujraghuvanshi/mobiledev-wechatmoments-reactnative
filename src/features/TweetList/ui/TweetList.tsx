import React, {ReactElement, useCallback, useEffect} from 'react';
import {RefreshControl, StyleSheet, View} from 'react-native';
import {FlatList} from 'react-native-gesture-handler';
import {connect} from 'react-redux';

import {BasicStyle, ITweet, RootState} from './../../../types';
import {Tweet} from './../../../features/Tweet/ui/Tweet';
import {useAppDispatch} from './../../../hooks';
import {fetchUserTweets} from './../../../features/TweetList/state/tweets.thunk';
import {
  loadMoreTweets,
  refreshTweets,
} from './../../../features/TweetList/state/tweets.slice';

interface ITweetListProps {
  tweets: Array<ITweet>;
  isRefreshing: boolean;
}

function TweetListComponent({
  tweets,
  isRefreshing,
}: ITweetListProps): ReactElement {
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(fetchUserTweets());
  }, [dispatch]);

  const handleLoadMore = useCallback(() => {
    dispatch(loadMoreTweets());
  }, [dispatch]);

  const refreshTweetsList = useCallback(() => {
    dispatch(refreshTweets());
  }, [dispatch]);

  return (
    <View style={styles.container}>
      <FlatList
        data={tweets}
        style={styles.list}
        renderItem={tweet => <Tweet tweet={tweet.item} />}
        onEndReached={handleLoadMore}
        onEndReachedThreshold={0.08}
        refreshControl={
          <RefreshControl
            refreshing={isRefreshing}
            onRefresh={refreshTweetsList}
          />
        }
      />
    </View>
  );
}

const mapStateToProps = (state: RootState) =>
  ({
    tweets: state.tweets.displayedTweets,
    isRefreshing: state.tweets.isRefreshing,
  } as ITweetListProps);

export const TweetList = connect(mapStateToProps)(TweetListComponent);

const styles: Partial<BasicStyle> = StyleSheet.create<Partial<BasicStyle>>({
  container: {
    flex: 1,
    paddingBottom: 8,
    paddingHorizontal: 8,
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
  },
  list: {
    width: '100%',
  },
});
