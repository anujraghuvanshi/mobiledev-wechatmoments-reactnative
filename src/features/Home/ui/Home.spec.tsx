import React from 'react';
import renderer from 'react-test-renderer';
import Home from './Home';
import {Header} from '../../Header/ui/Header';
import {TweetList} from '../../TweetList/ui/TweetList';

jest.mock('../../Header/ui/Header', () => ({
  Header: 'Header',
}));

jest.mock('../../TweetList/ui/TweetList', () => ({
  TweetList: 'TweetList',
}));

describe('Home Component', () => {
  it('should render component', () => {
    const tree = renderer.create(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('contains the correct testID', () => {
    const instance = renderer.create(<Home />).root;
    const viewWithTestID = instance.findByProps({testID: 'home-page'});
    expect(viewWithTestID).toBeTruthy();
  });

  it('renders Header and TweetList components', () => {
    const testRenderer = renderer.create(<Home />).root;
    const headerComponent = testRenderer.findByType(Header);
    const tweetListComponent = testRenderer.findByType(TweetList);

    expect(headerComponent).toBeTruthy();
    expect(tweetListComponent).toBeTruthy();
  });
});
