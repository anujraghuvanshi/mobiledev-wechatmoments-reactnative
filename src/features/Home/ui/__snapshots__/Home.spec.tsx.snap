// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component should render component 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
  testID="home-page"
>
  <Header />
  <TweetList />
</View>
`;
